samplers/aws/xray/sampling_rule.go

```diff
@@ -0,0 +1,79 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package xray
+
+import "sync"
+
+// ToDo: other fields will be used in business logic for remote sampling
+// centralizedRule represents a centralized sampling rule
+type centralizedRule struct {
+	// Centralized reservoir for keeping track of reservoir usage
+	reservoir *centralizedReservoir
+
+	// sampling rule properties
+	ruleProperties *ruleProperties
+
+	// Number of requests matched against this rule
+	//matchedRequests int64
+	//
+	// Number of requests sampled using this rule
+	//sampledRequests int64
+	//
+	// Number of requests burrowed
+	//borrowedRequests int64
+
+	// Provides system time
+	clock Clock
+
+	// Provides random numbers
+	rand Rand
+
+	mu sync.RWMutex
+}
+
+// properties is the base set of properties that define a sampling rule.
+type ruleProperties struct {
+	RuleName      *string            `json:"RuleName"`
+	ServiceType   *string            `json:"ServiceType"`
+	ResourceARN   *string            `json:"ResourceARN"`
+	Attributes    map[string]*string `json:"Attributes"`
+	ServiceName   *string            `json:"ServiceName"`
+	Host          *string            `json:"Host"`
+	HTTPMethod    *string            `json:"HTTPMethod"`
+	URLPath       *string            `json:"URLPath"`
+	ReservoirSize *int64             `json:"ReservoirSize"`
+	FixedRate     *float64           `json:"FixedRate"`
+	Priority      *int64             `json:"Priority"`
+	Version       *int64             `json:"Version"`
+}
+
+type samplingRuleRecords struct {
+	SamplingRule *ruleProperties `json:"SamplingRule"`
+}
+
+// getSamplingRulesOutput is used to store parsed json sampling rules
+type getSamplingRulesOutput struct {
+	SamplingRuleRecords []*samplingRuleRecords `json:"SamplingRuleRecords"`
+}
+
+// updateRule updates the properties of the user-defined and default centralizedRule using the given
+// *properties.
+func (r *centralizedRule) updateRule(rule *ruleProperties) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
```

Comment:
I didn't look in detail but noticed there are many locks / unlocks throughout the code. This makes me suspect a design issue - we poll for a complete sampling configuration and can create it each time, so in principle you should only need one synchronized update to all the sampling state rather than updating each individual portion behind locks. It should make the code easier to reason about.

Reply:
I think the major intent of locks throughout the code is to make sure atomic access. Yeah We are writing/updating these values when we do synchronize update at every interval but at the same time we are also reading these values. So locks at each point would make these accesses safer I think.

Reply:
Yeah I'm saying you should be able to atomically refresh all the state in one operation rather than processing each rule individually. Don't think of it as fetching individual rules, fetch the entire manifest, build a new manifest, and then swap it in